src/java.base/share/classes/java/util/zip/GZIPHeaderBuilder.java

```diff
@@ -0,0 +1,361 @@
+        }
+        this.fileComment = fileComment;
+        flags |= GZIPHeaderData.FCOMMENT;
+        return this;
+    }
+
+    /**
+     * Enable/Disable the CRC calculation of GZIP file header.
+     *
+     * @param calculateHeaderCRC if {@code true} the header data contains the lower 16 bytes of header CRC
+     * @return {@code this}
+     *
+     * @since 18
+     */
+    public GZIPHeaderBuilder calculateHeaderCRC(boolean calculateHeaderCRC) {
+        if (calculateHeaderCRC) {
+            flags |= GZIPHeaderData.FHCRC;
+        }
+        return this;
+    }
+
+    /**
+     *  Generate the GZIP header data
+     * @return the {@code record} of GZIP header data.
+     *
+     * @throws ZipException         If extra field size is out of range.
+     *                              Or if extra filed data layout is incorrect.
+     * @throws IllegalArgumentException     If compress method is not {@code Deflater.DEFLATED}
+     * @since 18
+     */
+    public GZIPHeaderData build() throws IOException {
+        byte[] headerBytes = generateBytes(cm, extraFieldBytes, filename, fileComment);
+        return new GZIPHeaderData(cm, flags, extraFieldBytes, filename, fileComment, headerCrc16, headerBytes);
+    }
+
+    /**
+     * Creates GZIP header bytes with optional header members and compress method.
+     * Per RFC-1952:
+     *    The filename and fileComment member should be String in
+     *          LATIN-1 (ISO-8859-1) character set.
+     *
+     *    A compliant compressor must produce files with correct ID1,
+     *          ID2, CM, CRC32, and ISIZE, but may set all the other fields in
+     *          the fixed-length part of the header to default values (255 for
+     *          OS, 0 for all others).  The compressor must set all reserved
+     *          bits to zero.
+     *
+     *   The XFL (extra Flags) is set to zero and OS is set to {@code OS_UNKNOWN (=255)}.
+     *   The FTEXT flag is set to zero and MTIME is filled with 0.
+     *
+     * @param cm                    compress method,
+     *                              per RFC-1952, 0-7 are reserved, 8 denotes "deflate".
+     *                              at present only support {@code Deflater.DEFLATED}
+     *
+     * @param extraFieldBytes
+     *        The byte array of extra filed, the generated header would calculate the
+     *        byte[] size and fill it before the byte[] in header.
+     * @param filename              the original file name in ISO-8859-1 character set
+     * @param fileComment           the file comment in ISO_8859-1 character set.
+     *
+     * @return Bytes of header data generated.
+     *
+     * @throws ZipException         If extra field size is out of range.
+     *                              Or if extra filed data layout is incorrect.
+     * @throws IllegalArgumentException     If compress method is not {@code Deflater.DEFLATED}.
+     *
+     * @since 18
+     */
+    public byte[] generateBytes(byte cm,
```

Comment:
Is there any reason to leave this public? If this is just an implementation detail, this should be kept private as public apis are maintenance burdens that are risky to change (hence csrs)